go-controller/pkg/crd/userdefinednetwork/v1/types.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=userdefinednetworks,scope=Namespaced
+// +kubebuilder:singular=userdefinednetwork
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// UserDefinedNetwork describe network request for a Namespace.
+type UserDefinedNetwork struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Spec is immutable"
+	// +kubebuilder:validation:XValidation:rule="self.topology != 'LocalNet' || self.role == 'Secondary'", message="LocalNet topology is not supported for primary network"
+	// +kubebuilder:validation:XValidation:rule="self.topology != 'Layer3' || has(self.subnets) && size(self.subnets) > 0", message="Subnets is required for Layer3 topology"
+	// +kubebuilder:validation:XValidation:rule="self.topology != 'Layer2' || self.role != 'Primary' || has(self.subnets) && size(self.subnets) > 0", message="Subnets is required for Primary Layer2 topology"
```

Comment:
maybe check the contents are a CIDR, instead of checking if the attribute is passed around.

Reply:
I will do both!